core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -410,21 +422,78 @@ public void index(Index index) {
     private void innerIndex(Index index) throws IOException {
         try (Releasable ignored = acquireLock(index.uid())) {
             lastWriteNanos = index.startTime();
-            final long currentVersion;
+            final boolean useAddDocument;
```

Comment:
Nit: I think it will be easier to understand if we can call this `canHaveDuplicates` instead. It has more implications than just using AddDocument. Also, the boolean doesn't directly control the lucene code but rather indirectly through the `currentVersion` variable.